### PR TITLE
Add glass map

### LINKED ISSUE #163: Draw glass map
Glass map is useful for picking substitute glasses or glasses with complementary properties to correct aberrations.

### PR DESCRIPTION
**Pull request by BrianGun**
# Pull Request Template

## Description

Added a function to draw glass maps of glass catalogs

Fixes #163 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Visual verification of plot

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by friggog** [comment]
Maybe marker=:none or marker=:nothing? I remember plots arts being a real pain and often just not working at all...

File: src/GlassCat/utilities.jl [hunk 1]
```diff
@@ -318,3 +318,28 @@ function plot_indices(glass::AbstractGlass; polyfit::Bool = false, fiterror::Boo
...
+function drawglassmap(glasscatalog::Module; λ = 550nm, glassfontsize = 3, showprefixglasses = false)
+    wavelength = Float64(ustrip(uconvert(u"μm", λ)))
+    indices = Vector{Float64}(undef,0)
+    dispersions = Vector{Float64}(undef,0)
+    glassnames = Vector{String}(undef,0)
+
+    for name in names(glasscatalog)
+        glass = eval(:($glasscatalog.$name))
+        glassstring = String(name)
+        hasprefix = occursin("_",glassstring)
```
Comment (alfredclwong) [inline src/GlassCat/utilities.jl:333]: There are some glasses like NBFD15_W and FDS90_SG in HOYA. Should these count as having prefixes? I'm not familiar enough with naming conventions to know.

File: src/GlassCat/utilities.jl [hunk 2]
```diff
@@ -318,3 +318,28 @@ function plot_indices(glass::AbstractGlass; polyfit::Bool = false, fiterror::Boo
...
+        
+        if typeof(glass) !== Module && index(glass,wavelength) > 0 && index(glass,wavelength) < 3 && (showprefixglasses ? true : !hasprefix)
```
Comment (alfredclwong) [inline src/GlassCat/utilities.jl:335]: Can this be simplified or explained by assigning it to an intermediate variable? I think it's quite hard to understand at first.

File: src/GlassCat/utilities.jl (hunk 1 above)
Comment (BrianGun) [reply src/GlassCat/utilities.jl]: Not sure what a suffix like that means. I believe prefixes indicate different versions of the same glass having nearly identical optical properties but slightly different physical properties, e.g., lead free vs  leaded, etc. If you don't filter these out then you end up with many glasses mapping to the same point on the glass map.

**Comment by BrianGun** [comment]
I tried marker = :none and that didn't do anything. The markers still showed up.

> Maybe marker=:none or marker=nothing? I remember plots args being a real pain and often just not working at all...

File: src/GlassCat/utilities.jl (hunk 1 above)
Comment (friggog) [reply src/GlassCat/utilities.jl:341]: It’s possible that invalid characters get replaced with underscores when the AGF files are parsed

**Comment by BrianGun** [comment]
Are there any remaining questions? In the future we can change the behavior with respect to glasses that have _ in the name. Are there any blocking issues?

File: src/GlassCat/utilities.jl [hunk 3]
```diff
@@ -318,3 +318,42 @@ function plot_indices(glass::AbstractGlass; polyfit::Bool = false, fiterror::Boo
...
+        glass = eval(:($glasscatalog.$name))
+        glassstring = String(name)
+        hasprefix = occursin("_",glassstring)
+ 
+        if typeof(glass) !== Module && (minindex <= index(glass,wavelength) <= maxindex)
+            f(x) = index(glass,x)
+            g = x -> ForwardDiff.derivative(f, x);
+            dispersion = g(wavelength)
+
+            if (mindispersion <= dispersion <= maxdispersion) && (showprefixglasses ? true : !hasprefix)    #don't show glasses that have an _ in the name. This prevents cluttering the map with many glasses of similar (index,dispersion).
```
Comment (alfredclwong) [inline src/GlassCat/utilities.jl]: I have a strong preference for the regular boolean expression `(showprefixglasses || !hasprefix)` as opposed to the ternary expression `(showprefixglasses ? true : !hasprefix)`. Was there a reason for adding this back in?

File: src/GlassCat/utilities.jl [hunk 4]
```diff
@@ -318,3 +318,42 @@ function plot_indices(glass::AbstractGlass; polyfit::Bool = false, fiterror::Boo
...
+
+            if (mindispersion <= dispersion <= maxdispersion) && (showprefixglasses ? true : !hasprefix)    #don't show glasses that have an _ in the name. This prevents cluttering the map with many glasses of similar (index,dispersion).  
+                push!(indices,index(glass,wavelength))          
+                push!(dispersions, dispersion)
+                push!(glassnames,String(name))
+            end
+        end
+    end
+    series_annotations = Plots.series_annotations(glassnames, Plots.font(family = "Sans", pointsize = glassfontsize, color = RGB(0.0,0.0,.4)))
+   scatter(dispersions,indices, xaxis = "dispersion", yaxis = "index", series_annotations = series_annotations, markersize = .001, legends = :none, markeralpha = 0.0, markershape = :none, title = "Glass Catalog: $glasscatalog")
```
Comment (alfredclwong) [inline src/GlassCat/utilities.jl]: Are `markersize` and `markershape` still needed?

**Comment by alfredclwong** [comment]
> @friggog or @alfredclwong could you review and approve if there are no remaining issues?
> 
> Are there any remaining questions? In the future we can change the behavior with respect to glasses that have _ in the name. Are there any blocking issues?
> 
> set the alpha of the markers to zero so now they don't draw. It's a hack but seems to work.

@BrianGun I posted a few questions above. I think it's safe to remove `markersize` and `markershape`, but it might be worth running the code again to see if it still works the way you want it to.

File: src/GlassCat/utilities.jl (hunk 4 above)
Comment (BrianGun) [reply src/GlassCat/utilities.jl]: no

**Comment by BrianGun** [comment]
@alfredclwong everything looks fine. Could you approve?

**Comment by alfredclwong** [comment]
> @alfredclwong everything looks fine. However I cannot verify that the examples are rendering correctly since there is some obscure problem on my local machine that prevents any of the example images from being rendered. Could you verify the glassmap example shows up properly in the Examples docs and then approve?

I don't see any changes to the Examples docs on this branch. Are the changes pushed? If you're having problems with the docs locally, you should still be able to see them here https://microsoft.github.io/OpticSim.jl/previews/PR164/examples/.

**Comment by BrianGun** [comment]
@alfredclwong The changes aren't in the Examples docs. They are in the GlassCat function docs. The new function shows up there in the preview docs. Everything seems to be working. Could you approve?

**Review by alfredclwong — APPROVED** [review]

**Comment by alfredclwong** [comment]
> @alfredclwong The changes aren't in the Examples docs. They are in the GlassCat function docs. The new function shows up there in the preview docs. Everything seems to be working. Could you approve?

Ah I thought you meant that you'd put in an example of the plot somewhere. Looks good.

**Pull request merged**